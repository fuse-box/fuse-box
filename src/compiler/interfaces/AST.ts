import { IASTScope } from '../Visitor/Visitor';

export interface ASTNode {
  raw?: string;
  type: keyof typeof ASTType;

  $assign_pattern?: boolean;
  $fuse_classInitializers?: Array<ASTNode>;
  $fuse_class_declaration_visited?: boolean;
  $fuse_decoratorForClassIdentifier?: string;
  $fuse_decoratorsCaptured?: boolean;
  $fuse_decoratorsProcessed?: boolean;
  $fuse_visited?: boolean;
  $transpiled?: boolean;
  alternate?: ASTNode;
  base?: ASTNode;
  chain?: ASTNode;
  consequent?: ASTNode;
  decorators?: Array<ASTNode>;
  definite?: boolean;
  elementTypes?: Array<ASTNode>;
  elements?: Array<ASTNode>;
  expressions?: Array<ASTNode>;
  initializer?: ASTNode;
  literal?: ASTNode;
  moduleReference?: ASTNode;
  optional?: boolean;
  prefix?: boolean;
  returnType?: ASTNode;
  scope?: IASTScope;
  shorthand?: boolean;
  source?: ASTNode;
  sourceType?: string;
  specifiers?: Array<ASTNode>;
  test?: ASTNode;
  typeAnnotation?: ASTNode;
  typeName?: ASTNode;
  loc?: { end: { column: number; line: number }; start: { column: number; line: number } };

  imported?: ASTNode;
  members?: Array<ASTNode>;
  method?: boolean;

  JSXElement?: ASTNode;

  abstract?: boolean;
  accessibility?: string;
  argument?: ASTNode;
  arguments?: Array<ASTNode>;
  async?: boolean;
  attributes?: Array<ASTNode>;
  body?: ASTNode | Array<ASTNode>;
  callee?: ASTNode;
  children?: Array<ASTNode>;
  closingElement?: ASTNode;
  computed?: boolean;
  declaration?: ASTNode;
  declarations?: Array<ASTNode>;
  declare?: boolean;
  exported?: ASTNode;
  expression?: any;
  generator?: boolean;
  id?: ASTNode;
  implements?: Array<ASTNode>;
  init?: null | ASTNode;
  key?: ASTNode;
  kind?: string;
  left?: ASTNode;
  local?: ASTNode;
  name?: any;
  object?: ASTNode;
  openingElement?: ASTNode;
  operator?: string;
  parameter?: ASTNode;
  params?: Array<ASTNode>;
  properties?: Array<ASTNode>;
  property?: ASTNode;
  quasi?: ASTNode;
  quasis?: Array<ASTNode>;
  right?: ASTNode;
  static?: boolean;
  superClass?: ASTNode;
  tag?: ASTNode;
  tail?: ASTNode;
  value?: any;
}

export const ASTType = {
  EnumDeclaration: 'TSEnumDeclaration',
  EnumMember: 'TSEnumMember',

  ParameterProperty: 'TSParameterProperty',
  StringKeyword: 'TSStringKeyword',
  TypeAnnotation: 'TSTypeAnnotation',

  AbstractClassProperty: 'TSAbstractClassProperty',
  AbstractMethodDefinition: 'TSAbstractMethodDefinition',
  AsExpression: 'TSAsExpression',
  DeclareFunction: 'TSDeclareFunction',
  DeclareKeyword: 'TSDeclareKeyword',
  InterfaceDeclaration: 'TSInterfaceDeclaration',
  NonNullExpression: 'TSNonNullExpression',
  TypeAliasDeclaration: 'TSTypeAliasDeclaration',
  TypeAssertion: 'TSTypeAssertion',

  ArrayExpression: 'ArrayExpression',
  ArrayPattern: 'ArrayPattern',
  ArrowFunctionExpression: 'ArrowFunctionExpression',
  AssignmentExpression: 'AssignmentExpression',
  AssignmentPattern: 'AssignmentPattern',
  AwaitExpression: 'AwaitExpression',
  BigIntLiteral: 'BigIntLiteral',
  BinaryExpression: 'BinaryExpression',
  Block: 'Block',
  BlockStatement: 'BlockStatement',
  BreakStatement: 'BreakStatement',
  CallExpression: 'CallExpression',
  CatchClause: 'CatchClause',
  ClassBody: 'ClassBody',
  ClassDeclaration: 'ClassDeclaration',
  ClassExpression: 'ClassExpression',
  ClassProperty: 'ClassProperty',
  ConditionalExpression: 'ConditionalExpression',
  ContinueStatement: 'ContinueStatement',
  DebuggerStatement: 'DebuggerStatement',
  Decorator: 'Decorator',
  DoWhileStatement: 'DoWhileStatement',
  EmptyStatement: 'EmptyStatement',
  ExportAllDeclaration: 'ExportAllDeclaration',
  ExportDefaultDeclaration: 'ExportDefaultDeclaration',
  ExportNamedDeclaration: 'ExportNamedDeclaration',
  ExportSpecifier: 'ExportSpecifier',
  ExpressionStatement: 'ExpressionStatement',
  ForInStatement: 'ForInStatement',
  ForOfStatement: 'ForOfStatement',
  ForStatement: 'ForStatement',
  FunctionDeclaration: 'FunctionDeclaration',
  FunctionExpression: 'FunctionExpression',
  HTMLClose: 'HTMLClose',
  HTMLOpen: 'HTMLOpen',
  Identifier: 'Identifier',
  IfStatement: 'IfStatement',
  Import: 'Import',
  ImportDeclaration: 'ImportDeclaration',
  ImportDefaultSpecifier: 'ImportDefaultSpecifier',
  ImportNamespaceSpecifier: 'ImportNamespaceSpecifier',
  ImportSpecifier: 'ImportSpecifier',
  JSXAttribute: 'JSXAttribute',
  JSXClosingElement: 'JSXClosingElement',
  JSXClosingFragment: 'JSXClosingFragment',
  JSXElement: 'JSXElement',
  JSXEmptyExpression: 'JSXEmptyExpression',
  JSXExpressionContainer: 'JSXExpressionContainer',
  JSXFragment: 'JSXFragment',
  JSXIdentifier: 'JSXIdentifier',
  JSXMemberExpression: 'JSXMemberExpression',
  JSXOpeningElement: 'JSXOpeningElement',
  JSXOpeningFragment: 'JSXOpeningFragment',
  JSXSpreadAttribute: 'JSXSpreadAttribute',
  JSXSpreadChild: 'JSXSpreadChild',
  JSXText: 'JSXText',
  LabeledStatement: 'LabeledStatement',
  Line: 'Line',
  Literal: 'Literal',
  LogicalExpression: 'LogicalExpression',
  MemberExpression: 'MemberExpression',
  MetaProperty: 'MetaProperty',
  MethodDefinition: 'MethodDefinition',
  NewExpression: 'NewExpression',
  ObjectExpression: 'ObjectExpression',
  ObjectPattern: 'ObjectPattern',
  OptionalCallExpression: 'OptionalCallExpression',
  OptionalMemberExpression: 'OptionalMemberExpression',
  Program: 'Program',
  Property: 'Property',
  RestElement: 'RestElement',
  ReturnStatement: 'ReturnStatement',
  SequenceExpression: 'SequenceExpression',
  SpreadElement: 'SpreadElement',
  Super: 'Super',
  SwitchCase: 'SwitchCase',
  SwitchStatement: 'SwitchStatement',
  TaggedTemplateExpression: 'TaggedTemplateExpression',
  TemplateElement: 'TemplateElement',
  TemplateLiteral: 'TemplateLiteral',
  ThisExpression: 'ThisExpression',
  ThrowStatement: 'ThrowStatement',
  TryStatement: 'TryStatement',
  module: 'module',
  script: 'script',

  AbstractKeyword: 'AbstractKeyword',

  AnyKeyword: 'TSAnyKeyword',
  ArrayType: 'TSArrayType',

  AsyncKeyword: 'AsyncKeyword',
  BigIntKeyword: 'BigIntKeyword',
  BooleanKeyword: 'TSBooleanKeyword',
  CallSignatureDeclaration: 'CallSignatureDeclaration',
  ClassImplements: 'ClassImplements',
  ConditionalType: 'ConditionalType',
  ConstructSignatureDeclaration: 'ConstructSignatureDeclaration',
  ConstructorType: 'ConstructorType',
  EmptyBodyFunctionExpression: 'EmptyBodyFunctionExpression',
  ExportAssignment: 'ExportAssignment',

  ExportKeyword: 'ExportKeyword',
  ExternalModuleReference: 'ExternalModuleReference',
  FunctionType: 'TSFunctionType',
  ImportEqualsDeclaration: 'TSImportEqualsDeclaration',
  ImportExpression: 'ImportExpression',
  ImportType: 'ImportType',
  IndexSignature: 'IndexSignature',
  IndexedAccessType: 'IndexedAccessType',
  InferType: 'InferType',

  InterfaceBody: 'InterfaceBody',
  InterfaceHeritage: 'InterfaceHeritage',
  IntersectionType: 'IntersectionType',
  LiteralType: 'TSLiteralType',
  MappedType: 'MappedType',
  MethodSignature: 'MethodSignature',
  ModuleBlock: 'ModuleBlock',
  ModuleDeclaration: 'TSModuleDeclaration',
  NamespaceExportDeclaration: 'NamespaceExportDeclaration',
  NeverKeyword: 'TSNeverKeyword',

  NullKeyword: 'TSNullKeyword',
  NumberKeyword: 'TSNumberKeyword',
  ObjectKeyword: 'ObjectKeyword',
  OptionalType: 'OptionalType',

  ParenthesizedType: 'ParenthesizedType',
  PrivateKeyword: 'PrivateKeyword',
  PropertySignature: 'PropertySignature',
  ProtectedKeyword: 'ProtectedKeyword',
  PublicKeyword: 'PublicKeyword',
  QualifiedName: 'QualifiedName',
  ReadonlyKeyword: 'ReadonlyKeyword',
  RestType: 'RestType',
  StaticKeyword: 'StaticKeyword',

  SymbolKeyword: 'SymbolKeyword',
  ThisType: 'ThisType',
  TupleType: 'TSTupleType',

  JSDocNullableType: 'JSDocNullableType',
  JSDocUnknownType: 'JSDocUnknownType',
  TypeLiteral: 'TypeLiteral',
  TypeOperator: 'TypeOperator',
  TypeParameter: 'TypeParameter',
  TypeParameterDeclaration: 'TypeParameterDeclaration',
  TypeParameterInstantiation: 'TypeParameterInstantiation',
  TypePredicate: 'TypePredicate',
  TypeQuery: 'TypeQuery',
  TypeReference: 'TSTypeReference',
  UnaryExpression: 'UnaryExpression',
  UndefinedKeyword: 'TSUndefinedKeyword',
  UnionType: 'UnionType',
  UnknownKeyword: 'UnknownKeyword',
  UpdateExpression: 'UpdateExpression',
  VariableDeclaration: 'VariableDeclaration',
  VariableDeclarator: 'VariableDeclarator',
  VoidKeyword: 'TSVoidKeyword',
  WhileStatement: 'WhileStatement',
  WithStatement: 'WithStatement',
  YieldExpression: 'YieldExpression',
};
