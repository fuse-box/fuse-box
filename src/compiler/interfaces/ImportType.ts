export enum ImportType {
  REQUIRE,
  FROM,
  RAW_IMPORT,
  DYNAMIC,
}
