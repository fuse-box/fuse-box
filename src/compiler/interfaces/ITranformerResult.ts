import { ASTNode } from './AST';
import { ITransformerRequireStatementCollection } from './ITransformerRequireStatements';

export interface ITransformerResult {
  ast?: ASTNode;
  requireStatementCollection: ITransformerRequireStatementCollection;
}
