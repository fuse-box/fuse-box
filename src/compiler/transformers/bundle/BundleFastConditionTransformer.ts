import { ASTNode } from '../../interfaces/AST';
import { computeBoolean } from '../../static_compute/computeBoolean';
import { IVisit, IVisitorMod } from '../../Visitor/Visitor';

export interface IBundleFastConditionTransformer {
  env: { [key: string]: any };
  isServer?: boolean;
  isBrowser?: boolean;
}
export function BundleFastConditionUnwrapper(options: IBundleFastConditionTransformer) {
  return (visit: IVisit): IVisitorMod => {
    const { node } = visit;
    if (node.type === 'IfStatement') {
      let operator = node.test.operator;
      let rightValue;
      let target: ASTNode;
      if (node.test) {
        if (node.test.type === 'MemberExpression') {
          target = node.test;
        }
        const left = node.test.left;
        const right = node.test.right;
        if (right && right.type === 'Literal') {
          rightValue = right.value;
        }
        if (left) {
          if (left.type == 'MemberExpression') target = left;
        }
      }
      let replacement: ASTNode;
      let result: boolean;

      if (target) {
        // process.env.*
        if (
          target.object.type === 'MemberExpression' &&
          target.object.object.name === 'process' &&
          target.object.property.name === 'env'
        ) {
          if (rightValue && operator) {
            result = computeBoolean(options.env[target.property.name], operator, rightValue);
            if (result) replacement = node.consequent;
            else replacement = node.alternate;
          }
        }
        if (target.object.name === 'FuseBox') {
          if (target.property.name === 'isBrowser') {
            if (options.isBrowser) replacement = node.consequent;
            else replacement = node.alternate;
          }

          if (target.property.name === 'isServer') {
            if (options.isServer) replacement = node.consequent;
            else replacement = node.alternate;
          }
        }
      }

      if (replacement) {
        if (replacement.body) return { replaceWith: replacement.body };
        else if (replacement.type === 'ExpressionStatement') return { replaceWith: replacement };
      } else {
        return { removeNode: true };
      }
    }
    return;
  };
}
