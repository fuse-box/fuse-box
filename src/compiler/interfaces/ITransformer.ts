import { ICompilerOptions } from '../../compilerOptions/interfaces';
import { Context } from '../../core/context';
import { IModule } from '../../moduleResolver/module';
import { IProductionContext } from '../../production/ProductionContext';
import { IVisit, IVisitorMod } from '../Visitor/Visitor';
import { ASTNode } from './AST';
import { ImportType } from './ImportType';

export type ITranformerCallback = (visit: IVisit) => undefined | IVisitorMod | void;

export interface ITransformerVisitors {
  onEachNode?: ITranformerCallback;
  onTopLevelTraverse?: ITranformerCallback;
}

export interface IRequireStatementModuleOptions {
  breakDependantsCache?: boolean;
}
export interface ITransformerCommon {
  compilerOptions: ICompilerOptions;
  ctx: Context;
  module: IModule;
  onRequireCallExpression?: (
    importType: ImportType,
    node: ASTNode,
    moduleOptions?: IRequireStatementModuleOptions,
  ) => void;
}

export interface ITransformerProduction {
  ctx: Context;
  module: IModule;
  productionContext: IProductionContext;
}

export interface ITransformer {
  target?: {
    test?: RegExp;
    type?: 'js' | 'js_ts' | 'ts';
  };
  commonVisitors?: (props?: ITransformerCommon) => ITransformerVisitors;
  productionWarmupPhase?: (props?: ITransformerProduction) => ITransformerVisitors;
}
