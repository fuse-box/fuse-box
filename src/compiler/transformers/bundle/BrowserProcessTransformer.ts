import { IVisit, IVisitorMod } from '../../Visitor/Visitor';
import { ITransformerSharedOptions } from '../../interfaces/ITransformerSharedOptions';
import {
  isPropertyOrPropertyAccess,
  createASTFromObject,
  defineVariable,
  createRequireStatement,
} from '../../Visitor/helpers';
import { ImportType } from '../../interfaces/ImportType';

interface BrowserProcessTransformProps {
  env?: { [key: string]: string };
}
export type IBrowserProcessTransform = BrowserProcessTransformProps & ITransformerSharedOptions;
export function BrowserProcessTransformer(options: IBrowserProcessTransform) {
  options = options || {};

  let globalEnvInserted = false;
  let entireEnvInserted = false;
  return (visit: IVisit): IVisitorMod => {
    const { node, parent } = visit;

    const accessList = isPropertyOrPropertyAccess(node, parent, 'process');
    if (accessList) {
      const variableAmount = accessList.length;
      if (variableAmount === 3 && accessList[1] === 'env') {
        const keyName = accessList[2];

        if (options.env && options.env[keyName] !== undefined) {
          return { replaceWith: { type: 'Literal', value: options.env[keyName].toString() } };
        }
        return { replaceWith: { type: 'Identifier', value: 'undefined' } };
      }
      if (variableAmount === 2) {
        switch (accessList[1]) {
          case 'version':
            return { replaceWith: { type: 'Literal', value: process.version }, avoidReVisit: true };
          case 'versions':
            return { replaceWith: { type: 'ObjectExpression', properties: [] }, avoidReVisit: true };
          case 'title':
            return { replaceWith: { type: 'Literal', value: 'browser' }, avoidReVisit: true };
          case 'umask':
            return { replaceWith: { type: 'Literal', value: 0 }, avoidReVisit: true };
          case 'browser':
            return { replaceWith: { type: 'Literal', value: true }, avoidReVisit: true };
          case 'cwd':
            return { replaceWith: { type: 'Literal', value: './' }, avoidReVisit: true };
          case 'env':
            if (options.env) {
              const response: IVisitorMod = {};
              if (!globalEnvInserted) {
                response.prependToBody = [defineVariable('___env', createASTFromObject(options.env))];
                globalEnvInserted = true;
              }
              response.replaceWith = { type: 'Identifier', name: '___env' };
              return response;
            } else return { replaceWith: createASTFromObject({}), avoidReVisit: true };

          default:
            // inserting require("process")
            // because we cannot match a know variable
            if (!entireEnvInserted) {
              entireEnvInserted = true;
              const statement = createRequireStatement('process', 'process');
              if (options.onRequireCallExpression)
                options.onRequireCallExpression(ImportType.REQUIRE, statement.reqStatement);
              return {
                avoidReVisit: true,
                prependToBody: [statement.statement],
              };
            }
            break;
        }
      }
    }
    return;
  };
}
