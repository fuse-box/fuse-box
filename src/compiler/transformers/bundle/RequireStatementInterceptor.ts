import { ImportType } from '../../interfaces/ImportType';
import { ITransformerSharedOptions } from '../../interfaces/ITransformerSharedOptions';
import { IVisit, IVisitorMod } from '../../Visitor/Visitor';

export function RequireStatementInterceptor(options: ITransformerSharedOptions) {
  options = options || {};
  return (visit: IVisit): IVisitorMod => {
    const { node } = visit;
    if (!options.onRequireCallExpression) {
      return;
    }

    if (node.type === 'CallExpression' && node.callee.name === 'require' && !node['emitted']) {
      const locals = visit.scope && visit.scope.locals ? visit.scope.locals : {};
      if (!locals[node.name]) {
        options.onRequireCallExpression(ImportType.REQUIRE, node);
      }
    }
    return;
  };
}
