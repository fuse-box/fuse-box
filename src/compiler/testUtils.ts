import { generate } from './generator/generator';
import { ASTNode } from './interfaces/AST';
import { ImportType } from './interfaces/ImportType';
import { ITransformer } from './interfaces/ITransformer';
import { ITransformerRequireStatementCollection } from './interfaces/ITransformerRequireStatements';
import { parseTypeScript } from './parser';
import { createGlobalContext } from './program/GlobalContext';
import { transpileModule } from './program/transpileModule';
import { GlobalContextTransformer } from './transformers/GlobalContextTransformer';

function cleanupForTest(node) {
  delete node.loc;
  delete node.raw;
  delete node.range;
}
export function initCommonTransform(props: {
  jsx?: boolean;
  code: string;
  transformers: Array<ITransformer>;
  props?: any;
}) {
  const requireStatementCollection: ITransformerRequireStatementCollection = [];
  function onRequireCallExpression(importType: ImportType, statement: ASTNode) {
    // making sure we have haven't emitted the same property twice
    if (!statement['emitted']) {
      Object.defineProperty(statement, 'emitted', { enumerable: false, value: true });
      cleanupForTest(statement.arguments[0]);
      cleanupForTest(statement);
      cleanupForTest(statement.callee);

      requireStatementCollection.push({ importType, statement });
    }
  }

  const ast = parseTypeScript(props.code, { jsx: props.jsx });

  // const ast = buntis.parseTSModule(props.code, {
  //   directives: true,
  //   jsx: props.jsx,
  //   next: true,
  //   loc: true,
  // });
  const tranformers = [GlobalContextTransformer().commonVisitors(props.props)];
  for (const t of props.transformers) {
    if (t.commonVisitors) {
      tranformers.push(t.commonVisitors({ ...props.props, onRequireCallExpression }));
    }
  }
  transpileModule({
    ast: ast as ASTNode,
    globalContext: createGlobalContext(),
    transformers: tranformers,
  });
  const res = generate(ast, {});

  return { code: res, requireStatementCollection };
}
