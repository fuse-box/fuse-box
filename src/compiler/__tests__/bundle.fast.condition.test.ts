import { initCommonTransform } from '../testUtils';
import { BundleFastConditionUnwrapper } from '../transformers/bundle/BundleFastConditionTransformer';
import { BrowserProcessTransformer } from '../transformers/bundle/BrowserProcessTransformer';
import { RequireStatementInterceptor } from '../transformers/bundle/RequireStatementInterceptor';
import { ImportType } from '../interfaces/ImportType';

const testTranspile = (props: { code: string; NODE_ENV?: string; target?: string }) => {
  return initCommonTransform({
    props: {
      ctx: { config: { env: { NODE_ENV: props.NODE_ENV || 'development' }, target: props.target || 'browser' } },
    },
    transformers: [BundleFastConditionUnwrapper(), BrowserProcessTransformer(), RequireStatementInterceptor()],
    code: props.code,
  });
};

describe('Browser fast condition', () => {
  describe('Process env', () => {
    it('should emit require statement', () => {
      const res = testTranspile({
        NODE_ENV: 'production',
        code: `
        if ( process.env.NODE_ENV === "development"){
          require("./dev")
        } else {
          require("./prod")
        }
      `,
      });

      expect(res.requireStatementCollection).toEqual([
        {
          importType: ImportType.REQUIRE,
          statement: {
            type: 'CallExpression',
            optional: false,

            callee: { type: 'Identifier', name: 'require' },
            arguments: [{ type: 'Literal', value: './prod' }],
          },
        },
      ]);
    });
  });

  describe('Fast condition', () => {
    it('should unwrap isBrowser', () => {
      const res = testTranspile({
        NODE_ENV: 'production',
        target: 'browser',
        code: `
        if ( FuseBox.isBrowser){
          console.log("isBrowser")
        }
      `,
      });
      expect(res.code).toMatchSnapshot();
    });

    it('should unwrap isServer', () => {
      const res = testTranspile({
        NODE_ENV: 'production',
        target: 'server',
        code: `
        if ( FuseBox.isServer){
          console.log("isServer")
        }
      `,
      });

      expect(res.code).toMatchSnapshot();
    });
  });
});
