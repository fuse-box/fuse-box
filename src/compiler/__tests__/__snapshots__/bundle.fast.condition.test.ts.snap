// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Browser fast condition FuseBox.* should unwrap isBrowser 1`] = `
"console.log(\\"isBrowser\\");
"
`;

exports[`Browser fast condition FuseBox.* should unwrap isServer 1`] = `
"console.log(\\"isServer\\");
"
`;

exports[`Browser fast condition Process env should unwrap dev 1`] = `
"console.log(\\"dev\\");
"
`;

exports[`Browser fast condition Process env should unwrap dev with alternate 1`] = `
"console.log(\\"this is dev\\");
"
`;
